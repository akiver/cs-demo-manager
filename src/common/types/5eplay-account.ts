export type FiveEPlayAccount = {
  id: string;
  domainId: string;
  nickname: string;
  avatarUrl: string;
  isCurrent: boolean;
};
