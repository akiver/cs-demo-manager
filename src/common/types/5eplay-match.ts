import type { Game } from 'csdm/common/types/counter-strike';
import type { DownloadStatus } from 'csdm/common/types/download-status';

export type FiveEPlayPlayer = {
  id: string; // uuid retrieved from the player's "domain"
  domainId: string; // the ID used in the player's page URL
  uid: number; // unique int player's id, used to know the player's team
  name: string;
  avatarUrl: string;
  hasWon: boolean;
  killCount: number;
  assistCount: number;
  deathCount: number;
  headshotCount: number;
  headshotPercentage: number;
  killDeathRatio: number;
  killPerRound: number;
  kast: number;
  threeKillCount: number;
  fourKillCount: number;
  fiveKillCount: number;
  firstKillCount: number;
  firstDeathCount: number;
  bombDefusedCount: number;
  bombPlantedCount: number;
  averageDamagePerRound: number;
};

type FiveEPlayTeam = {
  name: string;
  score: number;
  firstHalfScore: number;
  secondHalfScore: number;
  playerIds: number[];
};

export type FiveEPlayMatch = {
  id: string;
  game: Game;
  date: string;
  durationInSeconds: number;
  demoUrl: string;
  mapName: string;
  url: string;
  players: FiveEPlayPlayer[];
  teams: FiveEPlayTeam[];
  downloadStatus: DownloadStatus;
};
