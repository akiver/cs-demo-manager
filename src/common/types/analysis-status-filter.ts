export const AnalysisStatusFilter = {
  All: 'all',
  NotInDatabase: 'not-in-database',
  Pending: 'pending',
  InProgress: 'in-progress',
  InDatabase: 'in-database',
  Error: 'error',
} as const;

export type AnalysisStatusFilter = (typeof AnalysisStatusFilter)[keyof typeof AnalysisStatusFilter];
