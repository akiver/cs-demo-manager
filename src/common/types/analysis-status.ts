export const AnalysisStatus = {
  Pending: 'pending',
  Analyzing: 'analyzing',
  AnalyzeError: 'analyze-error',
  AnalyzeSuccess: 'analyze-success',
  Inserting: 'inserting-match',
  InsertError: 'insert-error',
  InsertSuccess: 'insert-success',
} as const;

export type AnalysisStatus = (typeof AnalysisStatus)[keyof typeof AnalysisStatus];
