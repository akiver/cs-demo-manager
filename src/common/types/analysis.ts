import type { DemoSource } from 'csdm/common/types/counter-strike';
import type { AnalysisStatus } from './analysis-status';

export type Analysis = {
  demoPath: string;
  demoChecksum: string;
  mapName: string;
  source: DemoSource;
  addedAt: string;
  status: AnalysisStatus;
  output: string;
};
