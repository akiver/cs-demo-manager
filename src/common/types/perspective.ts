export const Perspective = {
  Player: 'player',
  Enemy: 'enemy',
} as const;

export type Perspective = (typeof Perspective)[keyof typeof Perspective];
