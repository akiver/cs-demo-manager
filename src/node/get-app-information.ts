import os from 'node:os';

export type AppInformation = {
  platform: NodeJS.Platform;
  arch: NodeJS.Architecture;
  osVersion: string;
  electronVersion: string;
  chromeVersion: string;
};

export function getAppInformation(): AppInformation {
  return {
    platform: process.platform,
    arch: process.arch,
    osVersion: os.release(),
    electronVersion: process.versions.electron,
    chromeVersion: process.versions.chrome,
  };
}
