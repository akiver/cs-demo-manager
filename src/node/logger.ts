/* eslint-disable @typescript-eslint/no-explicit-any */
import path from 'node:path';
import { format } from 'node:util';
import fs from 'fs-extra';
import { watch } from 'chokidar';
import { getAppFolderPath } from 'csdm/node/filesystem/get-app-folder-path';

export interface ILogger {
  log: (...data: any[]) => void;
  warn: (...data: any[]) => void;
  error: (...data: any[]) => void;
  getLogFilePath: () => string;
  clear: () => Promise<void>;
}

type Level = 'log' | 'warn' | 'error';

class Logger implements ILogger {
  private logFolderPath: string;
  private logFilePath: string;
  private fileStream: fs.WriteStream | null = null;
  private watcher: fs.FSWatcher | null = null;
  private cache: any[] = [];
  private shouldLogToGlobalConsole = true;

  public constructor() {
    this.logFolderPath = path.join(getAppFolderPath(), 'logs');
    this.logFilePath = path.join(this.logFolderPath, 'csdm.log');
    // We only call the global Console methods when we are in development build or in the Electron renderer process
    // for the following reasons:
    // - On Windows AND in production build it may actually block the server process when logging large chunks of data.
    // - The Electron main process and WebSocket server processes stdout are never visible in production since there
    // is no terminal / window involved.
    //
    // Note:
    // process.type => "renderer" in development build as we start it from an Electron window
    // process.type => undefined in production build
    // This flag has been added after noticing that on Windows, logging the insertion of many demos NOT already in the
    // database (tested with 256 demos) it blocks the WebSocket server process in production build.
    this.shouldLogToGlobalConsole = IS_DEV || process.type === 'renderer';
  }

  public log = (...data: any[]) => {
    this.logToConsole(data, 'log');
    this.writeLogToFile(data, 'log');
  };

  public warn = (...data: any[]) => {
    this.logToConsole(data, 'warn');
    this.writeLogToFile(data, 'warn');
  };

  public error = (...data: any[]) => {
    this.logToConsole(data, 'error');
    this.writeLogToFile(data, 'error');
  };

  public getLogFilePath = () => {
    return this.logFilePath;
  };

  public clear = async () => {
    const fileExists = await fs.pathExists(this.logFilePath);
    if (!fileExists) {
      return;
    }

    try {
      await fs.writeFile(this.logFilePath, '');
    } catch (error) {
      this.error('Failed to clear log file', error);
      throw error;
    }
  };

  private logToConsole = (data: any[], level: Level) => {
    this.shouldLogToGlobalConsole && console[level](...data);
  };

  private jsonStringifyReplacer = (key: any, value: any) => {
    if (value instanceof Error) {
      return value.stack;
    }

    if (!value) {
      return value;
    }

    if (typeof value.toJSON === 'function') {
      return value.toJSON();
    }

    if (typeof value === 'function') {
      return '[function] ' + value.toString();
    }

    if (typeof value === 'object' && value !== null) {
      // Avoid possible circular reference
      if (this.cache.includes(value)) {
        return;
      }

      this.cache.push(value);
    }

    return value;
  };

  private formatDataToWrite = (...data: any[]) => {
    const json = data.map((value) => {
      if (value === undefined) {
        return undefined;
      }

      const string = JSON.stringify(value, this.jsonStringifyReplacer, 2);
      this.cache = [];

      if (string === undefined) {
        return undefined;
      }

      return JSON.parse(string);
    });

    return format(...json);
  };

  private buildLogLine = (data: any[], level: Level) => {
    const prefix = `${new Date().toUTCString()} | ${level.toUpperCase()} | ${process.env.PROCESS_NAME} |`;

    return `${prefix} ${this.formatDataToWrite(...data)}\n`;
  };

  private onFileDeleted = () => {
    this.fileStream?.close();
    this.fileStream = null;
  };

  private ensureLogFileExists = async () => {
    await fs.ensureFile(this.logFilePath);
  };

  private clearLogsIfFileMaxSizeReached = async () => {
    const { size } = await fs.stat(this.logFilePath);
    const fileSizeInMegabytes = size / (1024 * 1024);
    if (fileSizeInMegabytes > 1) {
      await this.clear();
    }
  };

  private watchForFileDeletion = () => {
    this.watcher?.close();
    this.watcher = watch(this.logFolderPath).on('unlink', this.onFileDeleted);
  };

  private writeLogToFile = async (data: any[], level: Level) => {
    await this.ensureLogFileExists();
    await this.clearLogsIfFileMaxSizeReached();

    if (this.fileStream === null) {
      this.fileStream = fs.createWriteStream(this.logFilePath, { flags: 'a' });
      this.watchForFileDeletion();
    }

    const line = this.buildLogLine(data, level);
    this.fileStream.write(line);
  };
}

globalThis.logger = new Logger();
