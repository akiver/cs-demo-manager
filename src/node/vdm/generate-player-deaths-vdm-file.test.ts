import { fileURLToPath } from 'node:url';
import fs from 'node:fs/promises';
import { describe, expect, it } from 'vitest';
import { Perspective } from 'csdm/common/types/perspective';
import type { PlaybackMatch } from '../database/watch/get-match-playback';
import { generatePlayerDeathsVdmFile } from './generate-player-deaths-vdm-file';

describe(`Generate player's deaths VDM file`, () => {
  it(`should generate a VDM file that focus on player's deaths`, async () => {
    const demoPath = fileURLToPath(new URL('player-deaths.dem', import.meta.url));
    const fakeVictimSteamId = 'fakeVictimSteamId';
    const fakeKillerSteamId = 'fakeKillerSteamId';
    const data: PlaybackMatch = {
      checksum: 'checksum',
      steamId: '12345',
      demoPath: demoPath,
      tickrate: 1,
      tickCount: 202,
      kills: [
        {
          roundNumber: 1,
          tick: 10,
          victimSteamId: fakeVictimSteamId,
          killerSteamId: fakeKillerSteamId,
        },
        {
          roundNumber: 1,
          tick: 13,
          victimSteamId: fakeVictimSteamId,
          killerSteamId: fakeKillerSteamId,
        },
        {
          roundNumber: 1,
          tick: 50,
          victimSteamId: fakeVictimSteamId,
          killerSteamId: fakeKillerSteamId,
        },
        {
          roundNumber: 2,
          tick: 200,
          victimSteamId: fakeVictimSteamId,
          killerSteamId: fakeKillerSteamId,
        },
      ],
    };

    const vdm = await generatePlayerDeathsVdmFile({
      match: data,
      steamId: fakeVictimSteamId,
      perspective: Perspective.Player,
      beforeDelaySeconds: 2,
      nextDelaySeconds: 4,
    });

    const vdmContent = await fs.readFile(vdm.getVdmPath(), 'utf8');
    expect(vdmContent).toMatchSnapshot();
  });
});
