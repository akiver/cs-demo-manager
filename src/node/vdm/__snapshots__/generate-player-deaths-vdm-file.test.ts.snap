// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generate player's deaths VDM file > should generate a VDM file that focus on player's deaths 1`] = `
"
// Generated by CS Demo Manager
demoactions
{
  
  \\"1\\"
  {
    factory \\"SkipAhead\\"
    name \\"SkipAhead1\\"
    starttick \\"0\\"
    skiptotick \\"8\\"
  }
  \\"2\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec2\\"
    starttick \\"8\\"
    commands \\"spec_lock_to_accountid fakeVictimSteamId\\"
  }
  \\"3\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec3\\"
    starttick \\"8\\"
    commands \\"spec_player_by_accountid fakeVictimSteamId\\"
  }
  \\"4\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec4\\"
    starttick \\"12\\"
    commands \\"spec_lock_to_accountid fakeVictimSteamId\\"
  }
  \\"5\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec5\\"
    starttick \\"12\\"
    commands \\"spec_player_by_accountid fakeVictimSteamId\\"
  }
  \\"6\\"
  {
    factory \\"SkipAhead\\"
    name \\"SkipAhead6\\"
    starttick \\"17\\"
    skiptotick \\"48\\"
  }
  \\"7\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec7\\"
    starttick \\"48\\"
    commands \\"spec_lock_to_accountid fakeVictimSteamId\\"
  }
  \\"8\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec8\\"
    starttick \\"48\\"
    commands \\"spec_player_by_accountid fakeVictimSteamId\\"
  }
  \\"9\\"
  {
    factory \\"SkipAhead\\"
    name \\"SkipAhead9\\"
    starttick \\"54\\"
    skiptotick \\"198\\"
  }
  \\"10\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec10\\"
    starttick \\"198\\"
    commands \\"spec_lock_to_accountid fakeVictimSteamId\\"
  }
  \\"11\\"
  {
    factory \\"PlayCommands\\"
    name \\"spec11\\"
    starttick \\"198\\"
    commands \\"spec_player_by_accountid fakeVictimSteamId\\"
  }
  \\"12\\"
  {
    factory \\"TextMessageStart\\"
    name \\"TextMessageStart12\\"
    starttick \\"53\\"
    message \\"Skipping to the next death in round 2\\"
    font \\"Trebuchet24\\"
    fadein \\"0.500\\"
    fadeout \\"0.500\\"
    holdtime \\"5.000\\"
    fxtime \\"0.000\\"
    FADEINOUT \\"1\\"
    x \\"10.000000\\"
    y \\"10.000000\\"
    r1 \\"255\\"
    g1 \\"255\\"
    b1 \\"255\\"
    a1 \\"0\\"
    r2 \\"0\\"
    g2 \\"0\\"
    b2 \\"0\\"
    a2 \\"0\\"
  }
  \\"13\\"
  {
    factory \\"ScreenFadeStart\\"
    name \\"ScreenFadeStart13\\"
    starttick \\"53\\"
    duration \\"1.000\\"
    holdtime \\"1.000\\"
    FFADE_IN \\"1\\"
    FFADE_OUT \\"1\\"
    r \\"0\\"
    g \\"0\\"
    b \\"0\\"
    a \\"255\\"
  }
  \\"14\\"
  {
    factory \\"StopPlayback\\"
    name \\"stop14\\"
    starttick \\"202\\"
  }
}
"
`;
